src/v/io/io_queue.cc

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "io/io_queue.h"
+
+#include "io/logger.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/sleep.hh>
+
+namespace experimental::io {
+
+io_queue::io_queue(
+  persistence* storage,
+  std::filesystem::path path,
+  completion_callback_type complete) noexcept
+  : storage_(storage)
+  , path_(std::move(path))
+  , complete_(std::move(complete)) {}
+
+void io_queue::start() noexcept {
+    log.debug("Starting io queue for {}", path_);
+    dispatcher_ = dispatch();
+}
+
+seastar::future<> io_queue::stop() noexcept {
+    log.debug("Stopping io queue for {}", path_);
+    auto dispatcher = std::exchange(
+      dispatcher_, seastar::make_ready_future<>());
+    stop_ = true;
+    cond_.signal();
+    co_await std::move(dispatcher);
+    co_await gate_.close();
+}
+
+const std::filesystem::path& io_queue::path() const noexcept { return path_; }
+
+seastar::future<> io_queue::open() noexcept {
+    log.debug("Opening {}", path_);
+    assert(!opened());
+    assert(running_.empty());
+    file_ = co_await storage_->open(path_);
+    cond_.signal();
+}
+
+bool io_queue::opened() const noexcept { return file_ != nullptr; }
+
+/*
+ * closing the file always succeeds. once seastar::file::close is invoked the
+ * handle is not longer usable, so we swallow any errors here and release units
+ * to avoid blocking other queues as there isn't much else we can do.
+ *
+ * failure to close a file is quite rare outside of the normal cases like trying
+ * to close an unknown or already closed file handle. these cases won't occur
+ * using the seastar api since it prevents those mistakes. in any case, what we
+ * care about most when closing a file is that its file descriptor resource is
+ * freed in the kernel to avoid reaching an open file limit.
+ */
+seastar::future<> io_queue::close() noexcept {
+    log.debug("Closing {}", path_);
+    assert(opened());
+
+    // ensure all on-going ops have completed
+    auto units = co_await seastar::get_units(
+      ops_, seastar::semaphore::max_counter());
```

Comment:
It looks like this ensures only that current in-flight operation is completed. It is possible to close the file even if we have pages in the `_pending` list.

Reply:
This is a bit counter intuitive to me. If the `io_queue` is per file then closing the `io_queue` should probably guarantee that all pending reads and writes are completed. This behavior is similar to `close` syscall when buffered I/O is used. It also awaits until all buffered I/O is done and can return an error.

Reply:
Also, I'm wondering why this (and also `open`) is not done in the `dispatch` background loop? This could eliminate extra synchronization. Instead of acquiring units here you can push promise to the queue and await it.

Reply:
> It is possible to close the file even if we have pages in the _pending list.

This is by design. I/O submission occurs independent of open/close. The former is for the I/O path, and the later is for resource management of total number of open file descriptors. I think that part of the issue here is the open/close terminology. In particular, the `close` terminology isn't a great analogy for `close` of a posix file. In the next PR that implements the open file descriptor quota the use will be much clear. In that PR I'll try to come up with a better name than open/close.

> Also, I'm wondering why this (and also open) is not done in the dispatch background loop?

This is because the I/O loop is unaware of higher level resource management of file descriptors. In the common case an IO queue will be opened and remain open forever. It's only under pressure that we may start closing file descriptors to stay under the limit.